### PR TITLE
Support application wide application packages associations

### LINKED ISSUE #2412: LoRa Cloud DAS configuration per application with URL config
#### Summary
LoRaCloud users should be able to select their own DAS server URL as it will not always be `https://das.loracloud.com`

#### How do you propose to implement this?
Add a `das-server-url` option taking and URL input

#### Can you do this yourself and submit a Pull Request?
Seems difficult, but maybe if I am helped with the steps to do it

### LINKED ISSUE #2473: LoRaCloud DAS package should send join event to the DAS
#### Summary
LoRa Cloud API evolved in February so application package should call DAS on join event. It does not matter if it is on joinReq or joinAccept

#### Why do we need this?
It allows the DAS to reset context associated to a device

#### What is missing? What do you want to see?
As [per documentation](https://www.loracloud.com/documentation/device_management?url=v1.html#deliver-uplink-message), the message send to the DAS at `/api/v1/uplink/send` should be
```
{
  DEVEUI:  {
     "msgtype":   "joining"       // Required.
     "timestamp": FLOAT,         // Optional. RX timestamp in seconds UTC
     "dr":        UINT1,         // Optional. Data rate
     "freq":      UINT4,         // Optional. Frequency
  }
}
```

#### Can you do this yourself and submit a Pull Request?

No

### LINKED ISSUE #2474: LoRaCloud DAS package uplinks management change
#### Summary
As [per documentation](https://www.loracloud.com/documentation/device_management?url=v1.html#post--api-v1-uplink-send) DAS should be called for all uplinks, so DAS gets all downlinks opportunity. 
For all non device management traffic, payload should be removed to ensure data privacy

#### What is missing? What do you want to see?

For all traffic not on DAS port registered by the user when creating the application package association, the uplink should be forwarded to the DAS:

```
{
  DEVEUI:  {
    "msgtype":   "updf"    // Optional. 'updf' is implied if not present
    "fcnt":      UINT4,    // Required. Frame counter
    "port":      UINT1,    // Required. Port
    "payload":   HEX,      // Required. HEX string with valid modem payload if port = device management port
    "dr":        UINT1,    // Optional. Data rate.
    "freq":      UINT4,    // Optional. Frequency
    "timestamp": FLOAT     // Optional. RX timestamp in seconds, UTC
    "dn_mtu":    UINT4,    // Optional. Downlink MTU
  },
}
```

#### Can you do this yourself and submit a Pull Request?

Maybe, although I might require help to implement tests

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

Closes #2412
Closes #2473
Closes #2474

#### Changes

- Add application packages default associations. These are the application-wide equivalent of normal application packages associations. They allow users to setup an application package for the whole application instead of an per-device basis. 
- Application packages may now handle _unrelated_ traffic (traffic on a different FPort than the one they've been associated with). This allows application packages to schedule downlinks even when their FPort hasn't been used.
- The LoRaCloud DAS application package now merges application-specific and device-specific state. This means that a normal association (device-level) may be used to override settings imposed by the default association (application-level).
- Add support for custom LoRaCloud DAS server URLs. Can be configured using the package data (and override on a per-device level, should that ever be needed). Old configurations are backwards-compatible.
- LoRaCloud DAS is now informed of new device sessions on `JoinAccept`.
- Some additional improvements to the structure of the API used by LoRaCloud DAS (there have been some changes since the original implementation was introduced).

#### Notes for Reviewers

I've did my best w.r.t naming but I realize that having both `associations` and `default-associations` may be confusing. Any suggestions are more than welcome.

@rvolosatovs, @KrishnaIyer please check the API changes mainly, and do a sanity check only if time allows you to do so.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/integrations/application-packages/lora-cloud-device-management.md [hunk 1]
```diff
@@ -51,3 +51,39 @@ This will enable the package on FPort `200` of the device `dev1` of application
     }
 }
 ```
+
+## Using a custom server URL
```
Comment (johanstokking) [inline doc/content/integrations/application-packages/lora-cloud-device-management.md:55]: Title case, please fix other occurrences

File: doc/content/integrations/application-packages/lora-cloud-device-management.md [hunk 2]
```diff
@@ -51,3 +51,39 @@ This will enable the package on FPort `200` of the device `dev1` of application
...
+
+{{< cli-only >}}
+
+The package may be configured to use a custom server URL using the package data:
+
+```bash
+# Create a JSON formatted file containing the uplink token and the server URL
+$ echo '{ "token": "AQEAdqwV67...", "server_url": "https://app.example.com/" }' > package-data.json
```
Comment (johanstokking) [inline doc/content/integrations/application-packages/lora-cloud-device-management.md:63]: This seems the first occurrence of where a `server_url` key is mentioned.

Don't we need a specification of the package data JSON, to begin with? Noting what the required values are, what optional is and what is default?

File: pkg/applicationserver/io/packages/loradms/v1/api/uplinks_test.go [hunk 3]
```diff
@@ -96,3 +96,15 @@ func TestUplinks(t *testing.T) {
 			}
 		})
 }
+
+func toUint8(x uint8) *uint8 {
```
Comment (johanstokking) [inline pkg/applicationserver/io/packages/loradms/v1/api/uplinks_test.go:100]: nit; convention is `uint8Ptr()`

File: pkg/applicationserver/io/packages/loradms/v1/package.go [hunk 4]
```diff
@@ -44,57 +45,69 @@ func (p *DeviceManagementPackage) RegisterHandlers(s *runtime.ServeMux, conn *gr
...
+			Type:      objects.UplinkUplinkType,
+			FCnt:      toUint32(msg.GetFCnt()),
+			Port:      toUint8(uint8(msg.GetFPort())),
+			Payload:   objects.Hex(msg.FRMPayload),
+			DR:        toUint8(uint8(settings.DataRateIndex)),
+			Freq:      toUint32(uint32(settings.Frequency)),
+			Timestamp: toFloat64(float64(msg.ReceivedAt.UTC().Unix())),
+		}
+		if fPort != msg.FPort {
+			log.FromContext(ctx).Debug("Uplink received on unhandled FPort; drop payload")
```
Comment (johanstokking) [inline pkg/applicationserver/io/packages/loradms/v1/package.go:91]: It's not dropping it; it's masking it
Also it's not really an unhandled port but rather a non-associated port (?)

File: pkg/applicationserver/io/packages/loradms/v1/package.go [hunk 5]
```diff
@@ -129,14 +142,58 @@ func (p *DeviceManagementPackage) HandleUp(ctx context.Context, assoc *ttnpb.App
...
-	}
-	packages.RegisterPackage(p, packages.CreateApplicationPackage(
+	}, packages.CreateApplicationPackage(
 		func(server io.Server, registry packages.Registry) packages.ApplicationPackageHandler {
 			return &DeviceManagementPackage{server, registry}
 		},
 	))
 }
+
+func toUint8(x uint8) *uint8 {
```
Comment (johanstokking) [inline pkg/applicationserver/io/packages/loradms/v1/package.go:189]: Idem

File: pkg/applicationserver/io/packages/redis/registry.go [hunk 6]
```diff
@@ -255,7 +262,188 @@ func (r ApplicationPackagesRegistry) Set(ctx context.Context, ids ttnpb.Applicat
...
+func (r ApplicationPackagesRegistry) SetDefaultAssociation(ctx context.Context, ids ttnpb.ApplicationPackageDefaultAssociationIdentifiers, gets []string, f func(*ttnpb.ApplicationPackageDefaultAssociation) (*ttnpb.ApplicationPackageDefaultAssociation, []string, error)) (*ttnpb.ApplicationPackageDefaultAssociation, error) {
+	appUID := unique.ID(ctx, ids.ApplicationIdentifiers)
+	fPort := r.fPortStr(ids.FPort)
+	uidKey := r.uidKey(appUID)
+	associationkey := r.associationKey(appUID, fPort)
+
+	defer trace.StartRegion(ctx, "set application package default association by id").End()
+
+	var pb *ttnpb.ApplicationPackageDefaultAssociation
+	err := r.Redis.Watch(func(tx *redis.Tx) error {
```
Comment (rvolosatovs) [inline pkg/applicationserver/io/packages/redis/registry.go:331]: IDEM

File: pkg/applicationserver/io/packages/redis/registry.go [hunk 7]
```diff
@@ -255,7 +262,188 @@ func (r ApplicationPackagesRegistry) Set(ctx context.Context, ids ttnpb.Applicat
...
+	defer trace.StartRegion(ctx, "list application package default associations by application id").End()
+
+	err := r.Redis.Watch(func(tx *redis.Tx) (err error) {
+		opts := []ttnredis.FindProtosOption{ttnredis.FindProtosSorted(false)}
+
+		limit, offset := ttnredis.PaginationLimitAndOffsetFromContext(ctx)
+		if limit != 0 {
+			total, err := tx.SCard(uidKey).Result()
+			if err != nil {
+				return ttnredis.ConvertError(err)
```
Comment (rvolosatovs) [inline pkg/applicationserver/io/packages/redis/registry.go:297]: I think we should only convert the top-level error

File: pkg/applicationserver/io/packages/loradms/v1/api/uplinks_test.go [hunk 8]
```diff
@@ -54,12 +54,12 @@ func TestUplinks(t *testing.T) {
 						eui := objects.EUI{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08}
 						resp, err := u.Send(objects.DeviceUplinks{
 							eui: &objects.LoRaUplink{
-								FCnt:      toUint32(42),
-								Port:      toUint8(200),
+								FCnt:      toUint32Ptr(42),
```
Comment (johanstokking) [inline pkg/applicationserver/io/packages/loradms/v1/api/uplinks_test.go:57]: `uint32Ptr`

File: pkg/applicationserver/io/packages/loradms/v1/api/objects/objects.go [hunk 9]
```diff
@@ -79,14 +81,29 @@ func (d DeviceUplinkResponses) UnmarshalJSON(b []byte) error {
...
+	PositionSolution      *PositionSolution `json:"position_solution"`
+	PendingRequests       *PendingRequests  `json:"pending_requests"`
+	FullfiledRequests     []Request         `json:"fulfilled_requests"`
+	Downlink              *LoRaDnlink       `json:"dnlink"`
+	InfoFields            InfoFields        `json:"info_fields"`
+	LogMessages           []LogMessage      `json:"log_messages"`
+	UploadedFiles         []File            `json:"uploaded_files"`
+	UploadedStreamRecords []Stream          `json:"uploaded_stream_records"`
+	LastUplink            *LoRaUplink       `json:"last_uplink"`
 }
```
Comment (Oliv4945) [inline pkg/applicationserver/io/packages/loradms/v1/api/objects/objects.go:107]: That is a `DEVICEINFO` structure, not an `UPLINK_RESPONSE`  
https://www.loracloud.sm.tc/documentation/device_management?url=v1.html#uplinkresponse  
https://www.loracloud.sm.tc/documentation/device_management?url=v1.html#device-info

Comment (Oliv4945) [inline pkg/applicationserver/io/packages/loradms/v1/api/objects/objects.go:107]: @adriansmares : Some fields are not a part of the `UPLINK_RESPONSE` object defined [here](https://www.loracloud.com/documentation/device_management?url=v1.html#uplink-response):
- `UploadedFiles`
- `UploadedStreamRecords`
- `LastUplink`

Then you correctly spotted issues in the documentation:

- `StreamRecords` is returning a `STREAM_UPDATE` array, doc [here](https://www.loracloud.com/documentation/device_management?url=v1.html#stream-update)
- `FullfiledRequests` should be named `PendingRequests` 
- `PendingRequests` should be added does not return a `[pending_request]` array but a `[request]` array


`FullfiledRequests` example:
```
 "fulfilled_requests":[
        {
          "type":"GETINFO",
          "param":[
            "signal",
            "temp"
          ]
        }
      ],
```

`pending_request` return by the DAS without requests (empty `requests` array):
```
      "pending_requests":{
        "requests":[        ],
        "id":2,
        "updelay":20,
        "upcount":1
      },
```

`stream_records` example:
```
      "stream_records":[
        [
          0,
          "010203"
        ]
      ],
```

**Pull request merged**